### ISSUE TITLE
Jobs[] not working as expected

### ISSUE THREAD

**Issue by rob-c** [ganga-devs/ganga#286]
The SpecialNames test is failing at the moment on 6.1.17

https://ganga.web.cern.ch/ganga/release/6.1.17-pre/reports/latest/html/summary_Ganga.test.XMLRepository.html

I don't think this is worth holding the release up, but I suspect this could mean we have potentially some unfortunate problems with RegistrySlices which could plague us in 6.1.17

We should decide if we're planning to drop support for special names or to propagate them correctly as we used to in 6.1.16 and prior. (even if we weren't doing it correctly).

This potentially stops users from doing things such as jobs[5:8] which I'm surprised doesn't cause problems already.

**Comment by rob-c** [comment]
Actually simply performing jobs[5] seems to fail as well as shown by the error in https://ganga.web.cern.ch/ganga/release/6.1.17-pre/reports/latest/html/summary_Ganga.test.Internals.html

Should we perhaps consider elevating this to a 6.1.17 bug as:

```jobs(5)``` works but ```jobs[5]``` does not.

**Comment by milliams** [comment]
Those work fine for me. The thing that no longer works of `jobs[-1]` which I'm not sure was being used and was broken the way it was previously implemented anyway. Are you seeing errors in your testing when doing `jobs[2:5]` or `jobs[2]`?

**Comment by rob-c** [comment]
The second test I linked is failing with ```j = jobs[self.id]``` which seems like something the test shouldn't be failing with if it's been working until recent releases.

**Comment by rob-c** [comment]
I'd say calling it broken is a little harsh, but it could have been done better. I know a few users who like to make use of the ```-n``` notation for addressing jobs when doing something similar to tasks in their own very complex scripts.
(I know of a few really complex LHCb scripts which break without this working)

**Comment by rob-c** [comment]
All special names appear broken based upon.
```
[12:45:38]
Ganga In [2]: jobs[1224:1225]
Ganga Out [2]:
Registry Slice: jobs[1224:1225] (0 objects)
```

I would expect that this should __really__ throw an error or work an I'm personally a little concerned pushing out with this. (I know we have other bugs to address so I'm wanting others opinion as to how serious this is)

**Comment by egede** [comment]
@rob-c  Does it work for subjobs?

**Comment by milliams** [comment]
That second link is calling `jobs[self.id]` which is then raising "Could not find object #-1" which suggests that the job has an `id` of `-1` which is very strange.

**Comment by rob-c** [comment]
@egede I don't know I'll have a look
@milliams I'm wondering does this represent some fundamental change in the test or in Core we've either forgotten about or failed to update the test for? I would feel some investigation would be good

**Comment by rob-c** [comment]
@egede This is all broken for subjobs an throws errors.
I would have estimated that this is an afternoons work to fix but it is something that could easily turn into 2 days work.
If we release with this notation broken can we at least advertise the fact we know this is broken and being addressed in 6.1.18?

**Comment by milliams** [comment]
I think before there was some logic in `Registry` to account for negative indices. However, I think it makes more sense in `RegistrySlice.__getitem__`. Something like:
```python
if isinstance(x, int):
    try:
        if x >= 0:  # NEW
            return addProxy(self.objects[x])
        else:  # NEW
            new_id = self.objects.keys()[x]  # NEW
            return addProxy(self.objects[new_id])  # NEW
    except IndexError:
        raise RegistryIndexError('list index out of range')
```

However, I'm not seeing the broken behaviour of slicing in my local checkout. Do you have an example few lines based on a clean gangadir?

**Comment by egede** [comment]
@milliams @rob-c Yes, the slicing seems to work for me as well.  Regarding the `[] ` notation, it works for me exactly the same as `()`. This was certainly not the original design, but I am not sure how long it has been broken for. For me it seems OK to go ahead with the release.

**Comment by rob-c** [comment]
@milliams I'd prefer to see this in the registry slice as otherwise we duplicate the logic in SubJobXML and the Registry.
If we can have some nice static methods then great, but I suspect this to be more difficult to code-up.

**Comment by rob-c** [comment]
@egede I suspect it's been broken fairly recently so I'd prefer to advertise that it's going to be fixed in 6.1.18 unless there are any objections?

**Comment by egede** [comment]
@rob-c Yes, that is fine. We should just make sure that we actually have a discussion about what the expected behaviour actually is.

**Comment by rob-c** [comment]
@egede agreed.
I suspect the expectation of the code in RegistrySlice is for the more complex 'general' index notation allowing everything which is broken.

@milliams The example I printed above does not work and should select both jobs for the ```:``` notation but selects nothing, and the ```-1``` notation is failing in the jobs, either for the test being broken or the -ve index being broken (it doesn't work on subjobs at least when I tested)

I think a possible workaround is that people could use jobs.select(a,b) which should still work atm for selecting ranges.

**Comment by drmarkwslater** [comment]
Just for my 2 cents, it seems from the discussion that the jobs[-1] notation is broken (which I believe was explicitly taken out in 6.1.16 though I'm not sure how functional it was before that) but registry slicing in general is OK? In any case, now we are in a much more stable state with the registry, I think adding this functionality back would be great but having it broken in 6.1.17 is no great shakes.

**Comment by rob-c** [comment]
I'm concerned you say 'explicitly taken out'. I must have missed something because this changes the way that people can access jobs when writing complex scripts.
Was this a deliberate attempt to remove this way of accessing jobs?

**Comment by drmarkwslater** [comment]
The specific commit is 64ce8415003c7babdc3cee03a0d20239c80a5ef4 where @milliams greatly simplified the __getitem__ to avoid deadlocks. The code could probably be put back here (or in RegistrySlice) now we're in a better situation but at the time we were trying to minimise any chance of deadlocks and reverting back to 6.0.44 code in some places seemed like the best idea. As I say, for 6.1.18 (which will have to come out quite quickly to fix Tasks) we should be able to reintroduce this functionality and all should be well.

**Comment by rob-c** [comment]
@drmarkwslater I'm now very concerned by this, do the internal implementation indexes always explicitly == the id from the object in the registry?

**Comment by rob-c** [comment]
I'm bugged that there is no sanity checking to allow for the Registry to return the result of ```-1``` this is a trivial 1 liner and is just a comparison and re-assignment of the argument so can't cause deadlocks.
I'll prepare a PR against this for 6.1.18 but I suspect this is due to accessing the Registry dictionary as if it's a list.

**Comment by drmarkwslater** [comment]
@rob-c For the first comment, Yes I believe they do. What I would certainly say though is that there was no check pre-6.1 and we never saw the error shown post 6.1 either (or at least I didn't) so I don't feel we need it here. It might be worth trying to come up with a test case to be sure but the coverage of this code is probably pretty extensive already.

For the second point, I agree - this functionality should be trivial to add back in and won't on it's own cause deadlocks as we are now not locking the registry for `__getitem__`.

**Comment by rob-c** [comment]
Bumping this to a future release as this needs looking at separately but I don't think this should hold up 6.1.19

**Comment by rob-c** [comment]
I imagine this has been fixed by #706 I'll check the release tests after that PR has been merged to confirm.

**Comment by rob-c** [comment]
This has been fixed I think, I can make use of `jobs[10]` and `jobs[-1]` correctly
